File: 011/ContainerWithMostWater.java
/**
 * ContainerWithMostWater
 */
public class ContainerWithMostWater {
    public int maxArea(int[] height) {
        int l = 0, r = height.length - 1;
        int max = 0, h = 0;
        while (l < r) {
            h = Math.min(height[l], height[r]);
            max = Math.max(max, (r - l) * h);
            while (height[l] <= h && l < r) l++;
            while (height[r] <= h && l < r) r--;
        }
        return max;
    }

    public static void main(String[] args) {
        ContainerWithMostWater sol = new ContainerWithMostWater();
        System.out.println(sol.maxArea(new int[]{1, 2, 4, 3}));
        System.out.println(sol.maxArea(new int[]{1, 8, 6, 2, 5, 4, 8, 3, 7}));
    }
}

File: 011/ContainerWithMostWater.scala
object ContainerWithMostWater {
  def maxArea(height: Array[Int]): Int = {
    var l = 0
    var r = height.length - 1
    var area = 0
    var h = 0
    while (l < r) {
      h = height(l) min height(r)
      area = area max h * (r - l)
      while (height(l) <= h && l < r)
        l += 1
      while (height(r) <= h && l < r)
        r -= 1
    }
    area
  }

  def main(args: Array[String]): Unit = {
//    var height = Array(1, 2, 4, 3)
    var height = Array(1, 8, 6, 2, 5, 4, 8, 3, 7)
    println(maxArea(height))
  }

}
